astropy/io/fits/file.py

```diff
@@ -578,7 +597,7 @@ def _open_zipfile(self, fileobj, mode):
 
         if close:
             zfile.close()
-        self.compression = 'zip'
+        self._file.seek(0)
```

Comment:
The seek in `_open_filename/filelike` is necessary in all cases in order to reset the position of the file handle after attempting to read the magic value. However, zip files are handled differently than either `bzip` or `gzip` because it is necessary to read the first entry in the zip archive and write it to a new file. You'll notice in the code above that the last operation on the underlying file handle was a write. So in order for the HDU detection to work properly, it is necessary to seek back to the beginning of the new file. I'll add a comment to clarify.